File: backend/src/services/service.service.ts
import { PrismaClient } from '@prisma/client';
import { Service } from '../models/types';

const prisma = new PrismaClient();

export class ServiceService {
  static async createService(name: string, price: number, description?: string): Promise<Service> {
    try {
      const service = await prisma.services.create({
        data: {
          name,
          price,
          description: description || null,
          created_at: new Date(),
          updated_at: new Date()
        }
      });

      return {
        id: service.id,
        name: service.name,
        price: service.price || 0,
        description: service.description || undefined,
        createdAt: service.created_at || new Date(),
        updatedAt: service.updated_at || new Date()
      };
    } catch (error) {
      console.error('Create service error:', error);
      throw error;
    }
  }

  static async getAllServices(): Promise<Service[]> {
    try {
      const services = await prisma.services.findMany({
        orderBy: {
          created_at: 'desc'
        }
      });

      return services.map(service => ({
        id: service.id,
        name: service.name,
        price: service.price || 0,
        description: service.description || undefined,
        createdAt: service.created_at || new Date(),
        updatedAt: service.updated_at || new Date(),
        service_type_id: service.service_type_id ?? undefined // Corrige le type pour éviter null
      }));
    } catch (error) {
      console.error('Get all services error:', error);
      throw error;
    }
  }

  static async updateService(serviceId: string, name: string, price: number, description?: string, service_type_id?: string): Promise<Service> {
    try {
      const updateData: any = {
        name,
        price,
        description: description || null,
        updated_at: new Date()
      };
      if (service_type_id) {
        updateData.service_type_id = service_type_id;
      }
      const service = await prisma.services.update({
        where: { id: serviceId },
        data: updateData
      });

      return {
        id: service.id,
        name: service.name,
        price: service.price || 0,
        description: service.description || undefined,
        createdAt: service.created_at || new Date(),
        updatedAt: service.updated_at || new Date(),
        service_type_id: service.service_type_id ?? undefined
      };
    } catch (error) {
      console.error('Update service error:', error);
      throw error;
    }
  }

  static async deleteService(serviceId: string): Promise<void> {
    try {
      await prisma.services.delete({
        where: { id: serviceId }
      });
    } catch (error) {
      console.error('Delete service error:', error);
      throw error;
    }
  }
}

File: frontend/mobile/admin-dashboard/lib/services/service_service.dart
import '../services/api_service.dart';
import '../models/service.dart';

class ServiceService {
  static const String _baseUrl = '/api/services';
  static final ApiService _apiService = ApiService();

  static Future<List<Service>> getAllServices() async {
    try {
      // Correction: utiliser le bon endpoint backend
      final response = await _apiService.get('$_baseUrl/all');
      if (response.data != null && response.data['data'] != null) {
        final List<Service> services = (response.data['data'] as List)
            .map((item) => Service.fromJson(item))
            .toList();
        return services;
      }
      return [];
    } catch (e) {
      print('Error fetching services: $e');
      rethrow;
    }
  }

  static Future<Service> createService({
    required String name,
    required double price,
    String? description,
    String? typeId,
  }) async {
    try {
      // Correction: utiliser le bon endpoint backend
      final response = await _apiService.post('$_baseUrl/create', data: {
        'name': name,
        'price': price,
        if (description != null) 'description': description,
        if (typeId != null) 'service_type_id': typeId,
      });
      if (response.statusCode != 201 && response.statusCode != 200) {
        throw 'Erreur lors de la création du service: ${response.statusCode}';
      }
      return Service.fromJson(response.data['data']);
    } catch (e) {
      print('[ServiceService] Error creating service: $e');
      rethrow;
    }
  }

  static Future<Service> updateService({
    required String id,
    String? name,
    double? price,
    String? description,
    String? typeId,
  }) async {
    try {
      // Correction: utiliser le bon endpoint backend
      final response = await _apiService.patch('$_baseUrl/update/$id', data: {
        if (name != null) 'name': name,
        if (price != null) 'price': price,
        if (description != null) 'description': description,
        if (typeId != null) 'service_type_id': typeId,
      });
      return Service.fromJson(response.data['data']);
    } catch (e) {
      print('[ServiceService] Error updating service: $e');
      rethrow;
    }
  }

  static Future<void> deleteService(String id) async {
    try {
      // Correction: utiliser le bon endpoint backend
      await _apiService.delete('$_baseUrl/delete/$id');
    } catch (e) {
      print('[ServiceService] Error deleting service: $e');
      rethrow;
    }
  }
}
